File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/3006_find-beautiful-indices-in-the-given-array-i/solutions/python3/solution.py
class Solution:
    def beautifulIndices(self, s: str, a: str, b: str, k: int) -> list:
        n = len(s)
        len_a = len(a)
        len_b = len(b)
        
        # Find all start indices where substring a occurs
        indices_a = [i for i in range(n - len_a + 1) if s[i:i + len_a] == a]
        
        # Find all start indices where substring b occurs
        indices_b = [i for i in range(n - len_b + 1) if s[i:i + len_b] == b]
        
        # We will use two pointers to determine the validity of the distance
        j = 0
        beautiful_indices = []
        
        # For each index in indices_a, check if a valid index in indices_b exists within distance k
        for i in indices_a:
            # Move j to the leftmost index in indices_b that is within the range [i-k, i+k]
            while j < len(indices_b) and indices_b[j] < i - k:
                j += 1
            
            # Check if there is any j that satisfies the condition |i - j| <= k
            found = False
            for m in range(j, len(indices_b)):
                if abs(indices_b[m] - i) <= k:
                    beautiful_indices.append(i)
                    found = True
                    break
                if indices_b[m] > i + k:
                    break
        
        return beautiful_indices

File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/3006_find-beautiful-indices-in-the-given-array-i/solutions/java/solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.TreeSet;

class Solution {
    public List<Integer> beautifulIndices(String s, String a, String b, int k) {
        List<Integer> result = new ArrayList<>();
        TreeSet<Integer> bIndices = new TreeSet<>();

        int aLen = a.length();
        int bLen = b.length();
        int n = s.length();

        // Collect all indices where substring b is found
        for (int j = 0; j <= n - bLen; j++) {
            if (s.substring(j, j + bLen).equals(b)) {
                bIndices.add(j);
            }
        }

        // Check for each index where substring a is found if it is beautiful
        for (int i = 0; i <= n - aLen; i++) {
            if (s.substring(i, i + aLen).equals(a)) {
                // Check if there exists an index j such that |j - i| <= k
                Integer floor = bIndices.floor(i + k);
                Integer ceiling = bIndices.ceiling(i - k);
                if ((floor != null && floor >= i) || (ceiling != null && ceiling <= i + k)) {
                    result.add(i);
                }
            }
        }

        return result;
    }
}
